File: Python/dp/word_wrap_problem.py
# A Dynamic programming solution
# for Word Wrap Problem

# A utility function to print
# the solution
# l[] represents lengths of different
# words in input sequence. For example,
# l[] = {3, 2, 2, 5} is for a sentence
# like "aaa bb cc ddddd". n is size of
# l[] and M is line width (maximum no.
# of characters that can fit in a line)
INF = 2147483647


def printSolution(p, n):
    k = 0
    if p[n] == 1:
        k = 1
    else:
        k = printSolution(p, p[n] - 1) + 1
    print('Line number ', k, ': From word no. ',
          p[n], 'to ', n)
    return k


def solveWordWrap(l, n, M):
    # For simplicity, 1 extra space is
    # used in all below arrays

    # extras[i][j] will have number
    # of extra spaces if words from i
    # to j are put in a single line
    extras = [[0 for i in range(n + 1)]
              for i in range(n + 1)]

    # lc[i][j] will have cost of a line
    # which has words from i to j
    lc = [[0 for i in range(n + 1)]
          for i in range(n + 1)]

    # c[i] will have total cost of
    # optimal arrangement of words
    # from 1 to i
    c = [0 for i in range(n + 1)]

    # p[] is used to print the solution.
    p = [0 for i in range(n + 1)]

    # calculate extra spaces in a single
    # line. The value extra[i][j] indicates
    # extra spaces if words from word number
    # i to j are placed in a single line
    for i in range(n + 1):
        extras[i][i] = M - l[i - 1]
        for j in range(i + 1, n + 1):
            extras[i][j] = (extras[i][j - 1] -
                            l[j - 1] - 1)

        # Calculate line cost corresponding
    # to the above calculated extra
    # spaces. The value lc[i][j] indicates
    # cost of putting words from word number
    # i to j in a single line
    for i in range(n + 1):
        for j in range(i, n + 1):
            if extras[i][j] < 0:
                lc[i][j] = INF;
            elif j == n and extras[i][j] >= 0:
                lc[i][j] = 0
            else:
                lc[i][j] = (extras[i][j] *
                            extras[i][j])

            # Calculate minimum cost and find
    # minimum cost arrangement. The value
    # c[j] indicates optimized cost to
    # arrange words from word number 1 to j.
    c[0] = 0
    for j in range(1, n + 1):
        c[j] = INF
        for i in range(1, j + 1):
            if (c[i - 1] != INF and
                    lc[i][j] != INF and
                    ((c[i - 1] + lc[i][j]) < c[j])):
                c[j] = c[i - 1] + lc[i][j]
                p[j] = i
    printSolution(p, n)


# Driver Code
#l = [3, 2, 2, 5]
#n = len(l)
#M = 6
n = int(input("Entre the lenght of the array "))
import array as arr
l = []
for i in range(0, n):
    x = int(input())
    l.append(x)
M = int(input("Entre the length of the  alphabets that you want in one line"))

solveWordWrap(l, n, M)


# __________________________***___________________________
# Time complexity o(n^2)
# Space complexity o(n^2)

# Sample Input
#Entre the lenght of the array 5
#2
#2
#3
#3
#5
#Entre the length of the  alphabets that you want in one line6

# Sample output

#Line number  1 : From word no.  1 to  2
#Line number  2 : From word no.  3 to  3
#Line number  3 : From word no.  4 to  4
#Line number  4 : From word no.  5 to  5

File: Java/dp/WordWrapProblem.java
/*In word wrap problem we've given a sequence of words as input
We have to find the number of words that can be fitted in a single line
at a time. So, for achieving this, we put breaks in the given sequence such
that the printed document looks good. The word editors such as MS office,
Libre Office, etc. use line breaks to make the document look nice.

We also have to assume that our word length does not exceed the line size.
This problem can be solved by following greedy as well as dynamic approaches.
Below is the Dynamic approach for achieving this.*/
import java.util.*;
import java.lang.*;
import java.io.*;
public class WordWrapProblem
{
    static int returnWordWrap (int wordSize[], int n, int lineSize)
    {
        int spareSpace[][] = new int[n+1][n+1];
        int leastCost[] = new int[n+1];
        for (int i=1;i<=n;i++)
        {
            spareSpace[i][i] = lineSize - wordSize[i-1];
            for (int j=i+1;j<=n;j++)
                spareSpace[i][j] = spareSpace[i][j-1] - wordSize[j-1] - 1;
        }
        leastCost[0] = 0;
        for (int i = 1; i <= n; i++)
        {
            leastCost[i] = Integer.MAX_VALUE;
            for (int j = 1; j <= i; j++)
            {
                // For storing cost of storing words[i, j]
                int cost;
                /*If spareSpace required is negative, then we can't place
                words[i,j] in a single line, else if we placed our last
                word, then we don't consider spareSpace, else calculate
                cost as per given cost function.*/
                if(spareSpace[j][i]<0){
                    cost = Integer.MAX_VALUE;
                }
                else if(i==n && spareSpace[j][i]>=0){
                    cost = 0;
                }
                else{
                    cost = spareSpace[j][i]*spareSpace[j][i]*spareSpace[j][i];
                }

                if (leastCost[j-1] != Integer.MAX_VALUE && cost != Integer.MAX_VALUE
                        && (leastCost[j-1] + cost < leastCost[i])) {
                    leastCost[i] = leastCost[j - 1] + cost;
                }
            }
        }
        return leastCost[n];
    }

    public static void main(String args[])
    {   //Below is the input code for finding least cost for word wrap
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter number of test cases you want to perform: ");
        int k = sc.nextInt();
        while(k-- > 0) {
            System.out.print("Enter size for an array of word size: ");
            int n = sc.nextInt();
            int[] wordSize = new int[n];
            System.out.print("Enter word size array elements: ");
            for(int i=0;i<n;i++) {
                wordSize[i] = sc.nextInt();
            }
            System.out.print("Enter the line size: ");
            int lineSize = sc.nextInt();
            int sol = returnWordWrap(wordSize, n, lineSize);
            System.out.println(sol);
        }
    }
}
/*
Sample Input/Output:
Input: Enter number of test cases you want to perform: 2
         Enter size for an array of word size: 4
         Enter word size array elements: 1 1 1 1
         Enter the line size: 5
         Enter size for an array of word size: 4
         Enter word size array elements: 3 1 2 4
         Enter the line size: 6

Output: 0
        35

Time and space complexity for this approach:
Time Complexity: O(n^2)
Auxiliary Space: O(n^2)
*/
